### ISSUE TITLE
Cannot compile "meta" test cases.

### ISSUE THREAD

**Issue by charlieamer** [skypjack/entt#314]
I am trying to compile and fiddle around with `test/entt/meta/meta.cpp`, but unfortunately I cannot compile it, I am not entirely sure what is wrong. I downloaded the source code today (commit ad7ee0b) from master branch. I am using the following configuration:

cmake 3.15.2
visual studio 2019 compiler - x86
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview>cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28008 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.
```
If you need a bit more information please ask. This is full compiler output:
```
[build] Starting build
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.exe" --build c:/Users/USER/Downloads/entt-master/entt-master/build --config Debug --target meta -- /m /property:GenerateFullPaths=true
[build] Microsoft (R) Build Engine version 16.3.0-preview-19377-01+dd8019d9e for .NET Framework
[build] Copyright (C) Microsoft Corporation. All rights reserved.
[build] 
[build]   gtest.vcxproj -> C:\Users\USER\Downloads\entt-master\entt-master\build\lib\Debug\gtestd.lib
[build]   odr.vcxproj -> C:\Users\USER\Downloads\entt-master\entt-master\build\test\odr.dir\Debug\odr.lib
[build]   gtest_main.vcxproj -> C:\Users\USER\Downloads\entt-master\entt-master\build\lib\Debug\gtest_maind.lib
[build]   meta.cpp
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,41) : error C2672: 'to_meta_function_helper': no matching overloaded function found [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(90) : message : see reference to alias template instantiation 'entt::internal::meta_function_helper_t<intdata_type::* >' being compiled [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(601) : message : see reference to function template instantiation 'bool entt::internal::setter<false,Type,pointer-to-member(0x0)>(entt::meta_handle,entt::meta_any,entt::meta_any)' being compiled [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(601) : message :         with [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(601) : message :         [ [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(601) : message :             Type=data_type [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(601) : message :         ] [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\test\entt\meta\meta.cpp(179) : message : see reference to function template instantiation 'entt::meta_factory<data_type> entt::meta_factory<data_type>::data<pointer-to-member(0x0),entt::as_alias_t,std::pair<properties,int>>(const uint32_t,std::pair<properties,int> &&) noexcept' being compiled [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\test\entt\meta\meta.cpp(184) : message : see reference to function template instantiation 'entt::meta_factory<data_type> entt::meta_factory<data_type>::data<pointer-to-member(0x0),entt::as_alias_t,std::pair<properties,int>>(const uint32_t,std::pair<properties,int> &&) noexcept' being compiled [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'int data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'int data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'int data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'const int data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'const int data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'const int data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'int *' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'int *' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'int *' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'const int *' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'const int *' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'const int *' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'empty_type data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'empty_type data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'empty_type data_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'int (*)[3]' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'int (*)[3]' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'int (*)[3]' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'int (array_type::* )[3]' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'int (array_type::* )[3]' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'int (array_type::* )[3]' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'int an_abstract_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'int an_abstract_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'int an_abstract_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl *)(Args...))': could not deduce template argument for 'Ret (__cdecl *)(Args...)' from 'char another_abstract_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(64) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...) const> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...) const)': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...) const' from 'char another_abstract_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(59) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(68,1) : error C2784: 'entt::internal::meta_function_helper<Ret(Args...)> entt::internal::to_meta_function_helper(Ret (__cdecl Class::* )(Args...))': could not deduce template argument for 'Ret (__cdecl Class::* )(Args...)' from 'char another_abstract_type::* ' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] C:\Users\USER\Downloads\entt-master\entt-master\src\entt/meta/factory.hpp(54) : message : see declaration of 'entt::internal::to_meta_function_helper' [C:\Users\USER\Downloads\entt-master\entt-master\build\test\meta.vcxproj]
[build] Build finished with exit code 1
[main] Failed to prepare executable target with name 'undefined'
```

**Comment by skypjack** [comment]
If we consider that with VC 19.22.27905.0 there are no problems and you use version 19.23.28008, maybe a regression in VS?
Unfortunately I don't have an updated VS available, I can't even reproduce the error. I test it only through the CI. So either you can help me with this or I must wait for an update from appveyor.

Are you using C++17 or `lastest` for the language?

**Comment by charlieamer** [comment]
That is set in cmake ? I am not sure where exactly I can set that, here are my configuration variables.
![image](https://user-images.githubusercontent.com/5943275/64707659-9d9c2000-d4b3-11e9-88c7-063e6e786997.png)

**Comment by skypjack** [comment]
No, I mean [this](https://stackoverflow.com/a/46759740/4987285). You can set the standard to use between C++14, C++17 and `latest`. What's the one set when you receive the error?

Also, if you can try something for me (because I cannot reproduce the error, so it's hard to say if this solves the problem), could you add this to the `factory.hpp` file among the other specializations of `to_meta_function_helper` and compile again the tests?
```
template<typename Ret, typename Class>
constexpr meta_function_helper<Ret()>
to_meta_function_helper(Ret Class:: *);
```

Thanks you.

**Comment by charlieamer** [comment]
I added those lines to factory.hpp, but it didn't solve the problem completely. Here is log with new error: https://www.diffchecker.com/BJDnWBEs . It seems that there are less errors. (On the right is the new compile log). Unfortunately, I have to go home, and I'll be back with you tomorrow, and I will examine the problem a bit more then.

I have also checked that different I use different standards, but that didn't change anything in compiler log. (It did for C++14 though, but I am guessing this shouldn't compile on C++14 anyway). By default it used C++17 and not latest.

**Comment by skypjack** [comment]
It's indeed a regression in VS, it wrongly interprets the type of the member we are registering. :smile: 
I'd rather open an issue to Microsoft to be honest, there isn't much I can do here unfortunately.

**Comment by skypjack** [comment]
Hi, any news on this? I plan to close the issue because it's a regression of the compiler and the problem should rather be reported to their bug tracker. Is it fine for you?

**Comment by charlieamer** [comment]
Hello, sorry for slow reply. I am not really sure what to write them, as I don't understand the issue in good detail. Could you please open the ticket instead and close this one ? If you need to test something on my version of vs compiler you can contact me and I can help you with that.

**Comment by skypjack** [comment]
Wait a moment. I've just realized it's evaluating the wrong branch of an `if constexpr`. May it be that it parse and evaluate also the branches that are going to be discarded? Otherwise, it fails to evaluate the guard and this is a problem anyway.

Can I ask you to make a test for me? Checkout `EnTT` from `master` (or use v3.1.1) and replace the `setter` function within `factory.hpp` with these functions:
```
template<bool Const, typename Type, auto Data>
std::enable_if_t<std::is_function_v<std::remove_pointer_t<decltype(Data)>> || std::is_member_function_pointer_v<decltype(Data)>, bool>
setter([[maybe_unused]] meta_handle handle, [[maybe_unused]] meta_any index, [[maybe_unused]] meta_any value) {
    bool accepted = false;

    if constexpr(!Const) {
        using helper_type = meta_function_helper_t<decltype(Data)>;
        using data_type = std::tuple_element_t<!std::is_member_function_pointer_v<decltype(Data)>, typename helper_type::args_type>;
        static_assert(std::is_invocable_v<decltype(Data), Type &, data_type>);
        auto *clazz = meta_any{handle}.try_cast<Type>();
        auto *direct = value.try_cast<data_type>();

        if(clazz && (direct || value.convert<data_type>())) {
            std::invoke(Data, *clazz, direct ? *direct : value.cast<data_type>());
            accepted = true;
        }
    }

    return accepted;
}


template<bool Const, typename Type, auto Data>
std::enable_if_t<!std::is_function_v<std::remove_pointer_t<decltype(Data)>> && !std::is_member_function_pointer_v<decltype(Data)>, bool>
setter([[maybe_unused]] meta_handle handle, [[maybe_unused]] meta_any index, [[maybe_unused]] meta_any value) {
    bool accepted = false;

    if constexpr(!Const) {
        if constexpr(std::is_member_object_pointer_v<decltype(Data)>) {
            using data_type = std::remove_cv_t<std::remove_reference_t<decltype(std::declval<Type>().*Data)>>;
            static_assert(std::is_invocable_v<decltype(Data), Type *>);
            auto *clazz = meta_any{handle}.try_cast<Type>();

            if constexpr(std::is_array_v<data_type>) {
                using underlying_type = std::remove_extent_t<data_type>;
                auto *direct = value.try_cast<underlying_type>();
                auto *idx = index.try_cast<std::size_t>();

                if(clazz && idx && (direct || value.convert<underlying_type>())) {
                    std::invoke(Data, clazz)[*idx] = direct ? *direct : value.cast<underlying_type>();
                    accepted = true;
                }
            } else {
                auto *direct = value.try_cast<data_type>();

                if(clazz && (direct || value.convert<data_type>())) {
                    std::invoke(Data, clazz) = (direct ? *direct : value.cast<data_type>());
                    accepted = true;
                }
            }
        } else {
            static_assert(std::is_pointer_v<decltype(Data)>);
            using data_type = std::remove_cv_t<std::remove_reference_t<decltype(*Data)>>;

            if constexpr(std::is_array_v<data_type>) {
                using underlying_type = std::remove_extent_t<data_type>;
                auto *direct = value.try_cast<underlying_type>();
                auto *idx = index.try_cast<std::size_t>();

                if(idx && (direct || value.convert<underlying_type>())) {
                    (*Data)[*idx] = (direct ? *direct : value.cast<underlying_type>());
                    accepted = true;
                }
            } else {
                auto *direct = value.try_cast<data_type>();

                if(direct || value.convert<data_type>()) {
                    *Data = (direct ? *direct : value.cast<data_type>());
                    accepted = true;
                }
            }
        }
    }

    return accepted;
}
```
It just splits the `setter` in two parts and uses sfinae rather than an `if constexpr` to evaluate the first guard. This will tell us where is the error. Let me know if it compiler or if you receive an error. In the latter case, please copy & paste here the log.

Thank you very much.

**Comment by charlieamer** [comment]
After I replaced setter with those 2 functions, the meta test case compiled without any problems :) I have also run the tests and they reported that they are all ok. Do you think this was a compiler issue ?

**Comment by skypjack** [comment]
That's **definitely** a compiler issue. It evaluates the branch even if it's not going to take if because the guard of the `if constexpr` is false. Using the two functions above I've just _forced_ the compiler to discard one of the branches before it has the chance to evaluate it.

I'm reopening the issue. I'll close it when the change is on `master`. Thanks for your help.
May I ask you another test during the next days if I refine the snippet above? I wrote it in a couple of minutes just to give you something to test.

**Comment by charlieamer** [comment]
Sure, no problem. Can you also submit the issue to the Microsoft as well ? I am not sure if I could explain the problem correctly :)

**Comment by skypjack** [comment]
Yep, not a problem, but I need to pack a small example to give them the chance to reproduce it easily.
I'll make a try and post here a small snippet today or tomorrow. If you can try it and confirm that it doesn't compile, I can open the issue to MS and point out this one to give you credits. :+1:

**Comment by skypjack** [comment]
Finally, the CI started offering VS v19.23 and I made my tests directly.
I've found that adding the following line is enough to solve your problem:
```
constexpr void to_meta_function_helper(...);
```
You can test the `wip` branch if you're interested. :+1: 
It will land `master` soon and this issue will be closed automatically.

Thank you for helping.

**Comment by BwdYeti** [comment]
9b53230f8391b01dd432aef956c8adc91c3571f0 undid this fix (752163a5c061fa6b4288d8cf198b77c602fa4736), so VS19 no longer builds. It works again when re-added.

**Comment by skypjack** [comment]
Oh, damn, I've updated VS locally and it was updated on the CI machine as well.
I'm re-adding that line back in place. Thanks for pointing this out.
